### PR TITLE
run: Add pull policy `--pull` flag 

### LINKED ISSUE #2101: run: Implement `--pull` argument
The run command only pulls an image if it's not present on the node. This behavior can create some confusion as the image being run not always matches the remote one. This is the exact same issue docker folks faced some years ago -> https://github.com/moby/moby/issues/13331.

### Implementation Options

Add a `--pull` argument that can be `missing` (default), `never` or `always` (same as docker run). The handling of the `latest` tag needs to be taken into consideration, Kubernetes always tries to pull images with that tag by default (https://kubernetes.io/docs/concepts/containers/images/#image-pull-policy), I'm not sure what's the behavior of Docker or other container runtimes.

### PR DESCRIPTION
**Pull request by burak-ok**
# run: Add pull policy `--pull` flag 

- `--pull missing` (default) only pulls the image when it is missing
- `--pull always` Pulls always
- `--pull never` Never pulls

The pull policy behavior is the same as Docker, but not the same as for K8s. For K8s in the default case the pull policy is missing, except when the `latest` tag is used. Then it is always pulling

Fixes #2101 
Slack Discussion: https://kubernetes.slack.com/archives/CSYL75LF6/p1697722813839619

### CODE REVIEW

**Comment by mqasimsarfraz** [comment]
I tried it together with @claudiamarcubina and it seems that it isn't working as expected for pull policy `always`:

```bash
$ go run -exec "sudo -E" ./cmd/ig image pull ghcr.io/inspektor-gadget/gadget/trace_open:latest
$ go run -exec "sudo -E" ./cmd/ig image pull ghcr.io/inspektor-gadget/gadget/trace_exec:latest
$ go run -exec "sudo -E" ./cmd/ig image tag ghcr.io/inspektor-gadget/gadget/trace_exec:latest ghcr.io/inspektor-gadget/gadget/trace_open:latest
# ensure trace_open is actually trace_exec
$ go run -exec "sudo -E" ./cmd/ig run ghcr.io/inspektor-gadget/gadget/trace_exec:latest
$ go run -exec "sudo -E" ./cmd/ig run ghcr.io/inspektor-gadget/gadget/trace_open:latest
# verify trace_open is trace_open
$ go run -exec "sudo -E" ./cmd/ig run --pull always ghcr.io/inspektor-gadget/gadget/trace_open:latest
```
It seems [oci-store](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/pkg/oci/oci.go#L148) used by `PullGadgetImage` isn't in sync with one used by [`GetGadgetImage`](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/pkg/oci/oci.go#L78)

**Comment by burak-ok** [comment]
> I tried it together with @claudiamarcubina and it seems that it isn't working as expected for pull policy `always`:

I tested it locally and its interesting:
```bash
➭ go run --exec "sudo -E" ./cmd/ig/... image list                                                                                   
INFO[0000] Experimental features enabled                
REPOSITORY                             TAG          DIGEST      
ghcr.io/burak-ok/trace_exec            latest       30507a96ca40
ghcr.io/burak-ok/trace_open            latest       d798a768d944

➭ go run --exec "sudo -E" ./cmd/ig/... image tag ghcr.io/burak-ok/trace_open ghcr.io/burak-ok/trace_exec
INFO[0000] Experimental features enabled                
Successfully tagged with ghcr.io/burak-ok/trace_exec:latest@sha256:d798a768d9449b13694aae2d0eb074e5f39cdddbc6a116fc26afe8a671afcfb9

➭ go run --exec "sudo -E" ./cmd/ig/... image list                                                       
INFO[0000] Experimental features enabled                
REPOSITORY                             TAG           DIGEST      
ghcr.io/burak-ok/trace_exec            latest        d798a768d944
ghcr.io/burak-ok/trace_open            latest        d798a768d944

➭ go run --exec "sudo -E" ./cmd/ig/... run ghcr.io/burak-ok/trace_exec --pull always                                                
INFO[0000] Experimental features enabled                
RUNTIME.CONTAINERNAME      PID           COMM             UID      GID      RET      FNAME                                              
minikube-docker            1894556888                     1041     57564    -1                                                          
minikube-docker            1902313085                     1041     57570    -1       -init                                              
^C%

➭ go run --exec "sudo -E" ./cmd/ig/... image list                                                       
INFO[0000] Experimental features enabled                
REPOSITORY                             TAG           DIGEST      
ghcr.io/burak-ok/trace_exec            latest        30507a96ca40
ghcr.io/burak-ok/trace_open            latest        d798a768d944

➭ go run --exec "sudo -E" ./cmd/ig/... run ghcr.io/burak-ok/trace_exec --pull never 
INFO[0000] Experimental features enabled                
RUNTIME.CONTAINERNAME                               PID                         PPID                        COMM             UID                         GID                        RETVAL                    
minikube-docker                                     64419                       19274                       bridge           0                           0                          0                         
minikube-docker                                     64425                       19274                       portmap          0                           0                          0                         
```

The first running gadget is clearly `trace_open`. But `image list` shows the correct `trace_exec` image right after the `ig run --pull always` command.
It can also be seen in the `ig run --pull never` command right after it that it is the correct `trace_exec` image.

Just wanted to share - Fixing it :)

**Comment by burak-ok** [comment]
> It seems [oci-store](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/pkg/oci/oci.go#L148) used by `PullGadgetImage` isn't in sync with one used by [`GetGadgetImage`](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/pkg/oci/oci.go#L78)

That was totally correct.  It didn't refer to the same `imageStore` instance. I thought it would always operate on the underlying filesystem - instead it caches it in memory in the instance somewhere.

I just had to "reload" the `imageStore`

File: pkg/oci/oci.go [hunk 1]
```diff
@@ -73,15 +77,38 @@ func getLocalOciStore() (*oci.Store, error) {
...
+	case PullImageAlways:
+		_, err := PullGadgetImage(ctx, image, authOpts)
+		if err != nil {
+			return nil, fmt.Errorf("pulling image %q: %w", image, err)
+		}
+		// We need to get the imageStore again because the previous call might have changed it
+		imageStore, err = getLocalOciStore()
+		if err != nil {
+			return nil, fmt.Errorf("getting local oci store: %w", err)
+		}
```
Comment (mauriciovasquezbernal) [inline pkg/oci/oci.go:98]: What about splitting `PullGadgetImage` into two functions, one of them `pullGadgetImage` receiving the OCI store, then here we can call it passing `imageStore` and avoid this logic.

Comment (burak-ok) [reply pkg/oci/oci.go]: Yes, that is the better way to resolve that

File: pkg/oci/oci.go [hunk 2]
```diff
@@ -143,13 +165,18 @@ func GetMetadata(ctx context.Context, image string, authOpts *AuthOptions) ([]by
...
 func PullGadgetImage(ctx context.Context, image string, authOpts *AuthOptions) (*GadgetImageDesc, error) {
 	ociStore, err := getLocalOciStore()
 	if err != nil {
 		return nil, fmt.Errorf("getting oci store: %w", err)
 	}
 
+	return pullGadgetImageToStore(ctx, ociStore, image, authOpts)
+}
+
+// pullGadgetImage pulls the gadget image into the given store and returns its descriptor.
```
Comment (claudiamarcubina) [inline pkg/oci/oci.go:178]:
```suggestion
// pullGadgetImageToStore pulls the gadget image into the given store and returns its descriptor.
```

**Review by claudiamarcubina — COMMENTED** [review]
Tested and works as described ✅ Thanks!

File: pkg/oci/oci.go [hunk 3]
```diff
@@ -188,11 +188,11 @@ func pullIfNotExist(ctx context.Context, imageStore oras.Target, authOpts *AuthO
 		return fmt.Errorf("resolving image %q: %w", image, err)
 	}
 
-	repo, err := NewRepository(image, authOpts)
+	repo, err := NewRepository(targetImage.String(), authOpts)
```
Comment (mauriciovasquezbernal) [inline pkg/oci/oci.go:218]: Do we need this change? 

If yes, should we change other places as well?

File: pkg/oci/oci.go [hunk 4]
```diff
@@ -73,15 +77,33 @@ func getLocalOciStore() (*oci.Store, error) {
 	return oci.New(defaultOciStore)
 }
 
-// GetGadgetImage pulls the gadget image and returns the a structure representing it.
-func GetGadgetImage(ctx context.Context, image string, authOpts *AuthOptions) (*GadgetImage, error) {
+// GetGadgetImage pulls the gadget image according to the pull policy and returns
+// a GadgetImage structure representing it.
+func GetGadgetImage(ctx context.Context, image string, authOpts *AuthOptions, pullPolicy string) (*GadgetImage, error) {
```
Comment (mauriciovasquezbernal) [inline pkg/oci/oci.go:82]: I was thinking if we should do the same for `GetEbpfObject` and `GetMetadata`, but they aren't used anymore. Perhaps you can remove them in this PR too?

**Review by mauriciovasquezbernal — COMMENTED** [review]
Works fine. Just one blocker question on the first commit. LGTM otherwise.

File: pkg/oci/oci.go (hunk 3 above)
Comment (burak-ok) [reply pkg/oci/oci.go:218]: Yes, we need it and probably other places too
I've tried with `ghcr.io/burak-ok/trace_exec` and that way it didn't add automatically the `:latest` tag, etc...

I am currently in favor of making the normalized image a type to mitigate further issues where we forgot to "normalize" the image

Comment (mauriciovasquezbernal) [reply pkg/oci/oci.go:180]: There are two changes on this commit, I agree the change below has to be done, but I'm wondering about this on the `NewRepository` line. I tried the following:

on main (121d76f3fea2c87527d595eac70156cb53cbdede)

```bash
$ sudo -E ig run  ghcr.io/burak-ok/trace_exec
INFO[0000] Experimental features enabled                
Error: getting gadget info: getting gadget image: pulling image "ghcr.io/burak-ok/trace_exec": downloading to local repository: failed to resolve ghcr.io/burak-ok/trace_exec: invalid reference
```

Performed the following change

```diff
$ git diff
diff --git pkg/oci/oci.go pkg/oci/oci.go
index 934a23f2..fa0761ac 100644
--- pkg/oci/oci.go
+++ pkg/oci/oci.go
@@ -192,7 +192,7 @@ func pullIfNotExist(ctx context.Context, imageStore oras.Target, authOpts *AuthO
        if err != nil {
                return fmt.Errorf("creating remote repository: %w", err)
        }
-       _, err = oras.Copy(ctx, repo, image, imageStore, image, oras.DefaultCopyOptions)
+       _, err = oras.Copy(ctx, repo, targetImage.String(), imageStore, targetImage.String(), oras.DefaultCopyOptions)
        if err != nil {
                return fmt.Errorf("downloading to local repository: %w", err)
        }

```

And then it worked fine:

```bash
$ sudo -E ig run  ghcr.io/burak-ok/trace_exec
INFO[0000] Experimental features enabled                
RUNTIME.CONTAINERNAME                  PID                  PPID                 COMM             UID                  GID                 RETVAL   
```

So probably we only need this change?

Comment (burak-ok) [reply pkg/oci/oci.go:180]: We need that also in other parts of it:
```bash
➭ go run --exec "sudo -E" ./cmd/ig/... image push burakokms/trace_exec
INFO[0000] Experimental features enabled                
Pushing burakokms/trace_exec...
Error: pushing gadget: copying to remote repository: Head "https://burakokms/v2/trace_exec/manifests/sha256:30507a96ca406d5453ca99b7ed8b51cfb09c1a54bb0127b56c155c8f68ba7b49": dial tcp: lookup burakokms: no such host
exit status 1

# (I didn't log in but the registry is the right one)
➭ go run --exec "sudo -E" ./cmd/ig/... image push burakokms/trace_exec
INFO[0000] Experimental features enabled                
Pushing burakokms/trace_exec...
Error: pushing gadget: copying to remote repository: HEAD "https://registry-1.docker.io/v2/burakokms/trace_exec/manifests/sha256:30507a96ca406d5453ca99b7ed8b51cfb09c1a54bb0127b56c155c8f68ba7b49": response status code 401: Unauthorized
exit status 1
```

Do we want to support this case without the full url?

Comment (burak-ok) [reply pkg/oci/oci.go]: For now I removed the line in `NewRepository`.

The discussion around if we want to have and support "default registries" has to be done before changing this

Comment (burak-ok) [reply pkg/oci/oci.go]: Discussion can be done in Issue https://github.com/inspektor-gadget/inspektor-gadget/issues/2190

**Pull request merged**